### PR TITLE
Limit libraries to versions that only support Python >=3.7, add support for Python 3.11

### PR DESCRIPTION
**Pull request by grahamalama**
In this PR, we: 
- Adjust Flask support to `>=2.0,<2.3`
- Adjust Sanic support to `>=21,<23`
- Add support for Django 4.2
- We also added tests to validate support for Python 3.11

### CODE REVIEW

**Comment by grahamalama** [comment]
@jwhitlock @relud, tagged y'all for review since you seemed to be here last. I'm still working on making changes to documentation, but I figured I'd tag you now to see if there was anything major I needed to fix.

File: docs/changelog.rst [hunk 1]
```diff
@@ -1,14 +1,14 @@
 Changelog
 ---------
 
-2022.8.0 (2022-08-18)
+2023.8.0
```
Comment (relud) [inline docs/changelog.rst:4]: why is 2022.8.0 being removed from the changelog?

Comment (grahamalama) [reply docs/changelog.rst]: Oof, that was certainly not intentional ([b6cd09e](https://github.com/mozilla-services/python-dockerflow/pull/76/commits/b6cd09e6f7a0b9046318e3e2ad8f878ad0f0d272))

**Comment by relud** [comment]
why aren't we dropping support for python 3.7 anymore? didn't it hit EoL yesterday?

> * Flask dropped Python 3.6 support in version 2.1.0 ([link](https://flask.palletsprojects.com/en/2.3.x/changes/#version-2-1-0)), so now we only test for compatibility with `Flask>=2.1`
> * Sanic dropped Python 3.6 support in version 21.3 ([link](https://sanic.dev/en/guide/release-notes/v21.3.html#python-3-7-only)), so now we only test for compatibility with `Sanic>=21`

This doesn't seem right. As an intermediate library normally I would expect to drop support for a library version in these cases:
 - when library version X reaches EoL
 - when library version X only runs on python<=3.Y and python 3.Y reaches EoL
 - when library version X is more than some number of release versions old (usually 2 or 3)
 - when a library doesn't have an EoL policy, but library's old version X hasn't received a backported security update this library considers critical

so for sanic, version 20.X works up to python 3.9 (which isn't EoL until 2025-10), is one of 3 major versions we support, but per sanic's security policy¹ version 20.12 reached EoL in 2022-12 and that's why we drop support for sanic<21.12.

¹ https://github.com/sanic-org/sanic/blob/main/SECURITY.md
> Sanic releases long term support release once a year in December. LTS releases receive bug and security updates for 24 months.

For flask, there is no EoL policy and version 0.12 works up to at least python 3.10. the 3 latest releases we support are 2.0, 2.1, and 2.2, so it might make sense to drop support for flask<2.0, but I don't see a reason to drop support for 2.0

File: tests/requirements/sanic.txt [hunk 2]
```diff
@@ -1,7 +1,6 @@
 # these are constrained by the files in tests/constraints/*.txt
 # to support a triple stack Django/Flask/Sanic
 aiohttp
-aioredis
```
Comment (relud) [inline tests/requirements/sanic.txt:4]: looks like this is actually still in use

File: tests/sanic/test_sanic.py [hunk 3]
```diff
@@ -62,8 +54,6 @@ def close(self):
 
 
 async def fake_redis(*args, **kw):
-    if aioredis.__version__.startswith("1."):
-        return FakeRedis1(*args, **kw)
```
Comment (relud) [inline tests/sanic/test_sanic.py:66]: looks like the definition of `FakeRedis1` should be dropped

File: AUTHORS.rst [hunk 4]
```diff
@@ -1,8 +1,9 @@
 - Peter Bengtsson (@peterbe)
+- Graham Beckley (@grahamalama)
 - Mike Cooper (@mythmon)
 - Will Kahn-Greene (@willkg)
 - Michael Kelly (@Osmose)
 - Jannis Leidel (@jezdez)
-- Les Orchard (@lmorchard)
 - Mathieu Leplatre (@leplatrem)
+- Les Orchard (@lmorchard)
```
Comment (jwhitlock) [inline AUTHORS.rst:8]: nit: unneeded change.

File: src/dockerflow/django/middleware.py [hunk 5]
```diff
@@ -3,28 +3,10 @@
...
-    from django.utils.deprecation import MiddlewareMixin
-except ImportError:  # pragma: no cover
-    MiddlewareMixin = object
-
-
-# Computed once, reused in every request
-_less_than_django_1_10 = VERSION < (1, 10)
-
-
-def is_authenticated(user):  # pragma: no cover
```
Comment (jwhitlock) [inline src/dockerflow/django/middleware.py:20]: Nice cleanup!

File: src/dockerflow/sanic/checks.py [hunk 6]
```diff
@@ -54,12 +54,9 @@ async def check_redis_connected(redis):
...
 
     """
-    import aioredis
+    from redis import asyncio as aioredis
 
-    if aioredis.__version__.startswith("1."):
-        RedisConnectionError = aioredis.ConnectionClosedError
-    else:
-        RedisConnectionError = aioredis.ConnectionError
+    RedisConnectionError = aioredis.ConnectionError
```
Comment (jwhitlock) [inline src/dockerflow/sanic/checks.py:59]: nit: could be further simplified, something like:

```python
from redis.asyncio import ConnectionError as RedisConnectionError
```

or even change `RedisConnectionError` in the except block below:

```python
    except ConnectionError as e:
        ...
 ```

File: tests/django/test_django.py [hunk 7]
```diff
@@ -13,16 +13,12 @@
 from django.db.utils import OperationalError, ProgrammingError
 from django.http import HttpResponse
 from django.test.utils import CaptureQueriesContext
+from django.utils.deprecation import MiddlewareMixin
 
 from dockerflow import health
 from dockerflow.django import checks
 from dockerflow.django.middleware import DockerflowMiddleware
 
-try:
```
Comment (jwhitlock) [inline tests/django/test_django.py:21]: Nice cleanup!

File: tests/flask/test_flask.py [hunk 8]
```diff
@@ -29,17 +29,18 @@ def load_user(user_id):
...
     login_manager = LoginManager(app)
     login_manager.user_loader(load_user)
     return app
 
 
-@pytest.fixture
-def app():
-    return create_app()
+@pytest.fixture()
+def client(app):
```
Comment (jwhitlock) [inline tests/flask/test_flask.py:42]: Nice!

File: tests/requirements/django.txt [hunk 9]
```diff
@@ -1,6 +1,3 @@
 django-redis
 pytest-django
-# these are constrained by the files in tests/constraints/*.txt
-# to support a triple stack Django/Flask/Sanic
-Django; python_version >= '3.0'
-Django<2.0; python_version < '3.0'
+Django
```
Comment (jwhitlock) [inline tests/requirements/django.txt:3]: 👍

File: tox.ini [hunk 10]
```diff
@@ -4,18 +4,18 @@ minversion = 1.8
 envlist =
     py38-lint
     py38-docs
-    py{37,38,39,310}-dj32
-    py{38,39,310}-dj{40,41}
-    py{37,38,39,310}-fl{012,10,11,20,21,22}
-    py{37,38,39}-s20
-    py{37,38,39,310}-s{21,22}
+    py{37,38,39,310,311}-dj32
+    py{38,39,310,311}-dj{40,41}
```
Comment (jwhitlock) [inline tox.ini:8]: nit: could add [Django 4.2](https://docs.djangoproject.com/en/4.2/releases/4.2/), the LTS version released in April 2023. However, that could be a PR after this one, and would require at least a few more lines in this file.

**Review by jwhitlock — DISMISSED** [review]
Thanks @grahamalama! The Django changes look correct. The Sanic and Flask ones look sane as well, however I don't claim to know those frameworks as well, so I'm relying on the passing tests for a 👍.

I think we could add Django 4.2 to the matrix as well, but my local tests show `pytest tests/core/ tests/django --nomigrations` fails with `pytest: error: unrecognized arguments: --nomigrations`, so some more special casing will be needed. I'd like to get Django 4.2 into the next release though.

**Update**: I had an error in `tox.ini`. If I install `-rtests/requirements/django.txt`, Django 4.2 is happy as well. I can write a follow-on PR to add Django 4.2.

**Comment by grahamalama** [comment]
> why aren't we dropping support for python 3.7 anymore? didn't it hit EoL yesterday?

@relud I considered this, but I noticed that Django 3.2 (which doesn't hit EoL until April of next year) still supports 3.7.

To your point about 3.7 hitting EoL recently, I figured we could maybe throw in some deprecation warnings (if we want) about dropping support for Python 3.7, then do so in April of next year when Django 3.2 hits _its_ EoL.

> For flask, there is no EoL policy and version 0.12 works up to at least python 3.10. the 3 latest releases we support are 2.0, 2.1, and 2.2, so it might make sense to drop support for flask<2.0, but I don't see a reason to drop support for 2.0

Good point -- I'll add Flask 2.0 back 👍🏻

File: src/dockerflow/sanic/checks.py (hunk 6 above)
Comment (grahamalama) [reply src/dockerflow/sanic/checks.py]: [b850e82](https://github.com/mozilla-services/python-dockerflow/pull/76/commits/b850e8202bf7a7ff15ff63b4b2f6199caedec3c2)

File: tox.ini (hunk 10 above)
Comment (grahamalama) [reply tox.ini]: This was pretty easy to pull in, so I went ahead and did it [3dccd82](https://github.com/mozilla-services/python-dockerflow/pull/76/commits/3dccd82419b75150a7e62dfb172fa602fac6b7c7)

**Review by relud — APPROVED** [review]

File: tests/constraints/django-4.2.txt [hunk 11]
```diff
@@ -0,0 +1 @@
+Django>=4.2,<4.3
```
Comment (jwhitlock) [inline tests/constraints/django-4.2.txt:1]: 😍

**Pull request merged**